docs: Custom Applications multiple permissions (#2806)
commercetools/merchant-center-application-kit: website/src/content/concepts/oauth-scopes-and-user-permissions.mdx

This grouping determines the **mapping and relation between OAuth Scopes and user permissions**.

## Permission groups

Every Custom Application has a **default permission group** with a pair of view and manage permissions. This group maps to the OAuth Scopes specified in the [oAuthScopes](/api-reference/application-config#oauthscopes) field of the Custom Application config.

> However, you might need more granular access control to fulfill specific business requirements. For example, suppose your Custom Application manages products, discounts, and orders. And you want a group of users to only manage products and discounts while another group handles orders.<br />

Reviewer: ```suggestion
However, you might need more granular access control to fulfill specific business requirements. Suppose your Custom Application manages products, discounts, and orders, and you want a group of users to only manage products and discounts while another group handles orders. <br />
```

In this kind of scenario a single permission group (the default one) is not enough as you would need multiple permission groups fulfil the access requirements.

> To enable such use cases, you can define additional permission groups that have different access requirements. See [additionalOAuthScopes](/api-reference/application-config#additionaloauthscopes) field in the Custom Application config. 

Reviewer 2: @Anshuman71 could you please rework your suggestion for the paragraph starting with: 'However, you might need more..'. 🙏 It is not so easy to read and there is a sentence starting with 'And'.

<Info>

This feature is available from version `21.21.0` onwards.

</Info>

In the following example, we are defining two additional permission groups. The group `delivery` allows users to manage incoming orders while the group `promotion` enables users to work on discount and promotion campaigns.

```json title="custom-application-config.json" highlightLines="6-17"
{
  "oAuthScopes": {
    "view": ["view_products", "view_customers"],
    "manage": ["manage_products"]
  },
  "additionalOAuthScopes": [
    {
      "name": "delivery",
      "view": [],
      "manage": ["manage_orders"]
    },
    {
      "name": "promotion",
      "view": [],
      "manage": ["manage_orders", "manage_discount_codes"]
    }
  ]
}
```

<Info>

The default permission group is always defined, even when adding additional groups.

When additional groups are defined, the default group can be left "empty," without specifying any OAuth Scopes.

> Even so, remember that the "view-only" user permission must be enabled to allow access to the Custom Application.

Reviewer 3: Hmm here we're actually mixing things up a bit. The first part about defining the permission groups is about the Custom Application configuration while the "view-only" part is about assigning team permissions.

Having the two points in one sentence makes is look like they are strictly related when configuring the app.

Maybe you can try rephrasing this again in case the original version wasn't clear enough?

> Even so, remember that the "view-only" user permission must be enabled to allow access to the Custom Application.

Reviewer: @emmenko, thanks for highlighting that. I've rephrased it for clarity.

This grouping determines the **mapping and relation between OAuth Scopes and user permissions**.

> # Additional OAuth Scopes

Reviewer 3: Wondering if we should maybe use the wording "permission group" here and have it as a subsection of `OAuth Scopes`.

Reviewer 3: Also, we should include the info about the min version to use the granular permission groups.

```
<Info>

This feature is available from version `21.21.0` onwards.

</Info>
```

> Defining `oAuthScopes` in the Custom Application config allows using permissions limited to 1 unique pair (view/manage) specific to the Custom Application.

Reviewer 3: Do you think it would help if we clarify the distinction between the default permission group and the additional groups?

> For more granular permissions, for example, to allow the team access to only certain parts or functionality of the Custom Application, [additional OAuth Scopes](https://docs.commercetools.com/api/scopes) can be requested as part of various permission groups. 

Reviewer 3: I think it would help if we can maybe provide a more "real-world" concrete example and how it can be solved by using multiple permission groups.

However, in certain cases more granular access control is needed. For example, if multiple teams accross the organization have access to a Custom Application but only one team is allowed to manage discount codes, the default permission pair is not enough.

For more granular permissions [additional OAuth Scopes](https://docs.commercetools.com/api/scopes) can be requested as part of various additional permission groups. 

These additional OAuth Scopes must be specified in your Custom Application config, using the [`additionalOAuthScopes` field](/api-reference/application-config#additionaloauthscopes).

In the following example, permission group named `team-a` allows to manage orders but not manage discount codes, while permission group named `team-b` allows both.

```json title="custom-application-config.json" highlightLines="6-17"
{
  "oAuthScopes": {
    "view": ["view_products", "view_customers"],
    "manage": ["manage_products"]
  },
  "additionalOAuthScopes": [
    {
      "name": "team-a",
      "view": [],
      "manage": ["manage_orders"]
    },
    {
      "name": "team-b",
      "view": [],
      "manage": ["manage_orders", "manage_discount_codes"]
    }
  ]
}
```

> 

Reviewer 3: ```suggestion

<Info>

The default permission group is always defined, even when adding additional groups.

When additional groups are defined, the default group can be left "empty", without specifying any OAuth Scopes.

Even so, remember that the "view-only" user permission must be enabled to allow access to the Custom Application.

</Info>

```

Author: 💎 
Just to double check:

> The default permission group is always defined, even when **not** adding additional groups

Reviewer 3: Hmm I believe is correct the way it is.

How I read it:

1. `The default permission group is always defined` implies that this is a fact and won't change.
2. `even when adding additional groups` implies that the previous statement `always defined` will continue to be true.

To counter the first statement you would use something like `except`.

Does that make sense 🤔

Author: Thanks for explaining, I get your point now 😄

<Info>

> The permission names are unique to each Custom Application and, by default, they derive from the `entryPointUriPath`, based on the following format: `{View,Manage}<EntryPointUriPath>`.

Reviewer 4: We have this sentence above:
> Each Custom Application gets a **unique pair** of user permissions: "view" and "manage."

I'm wondering if we should update that because a Custom Application using `additionalOAuthScopes` will not only get one pair of user permissions.

| `avengers-01` | `{View,Manage}Avengers/01` |
| `avengers_01` | `{View,Manage}Avengers_01` |

> Optional additional permissions derive from the `entryPointUriPath` and the provided permission group name, based on the following format: `{View,Manage}<EntryPointUriPath><PermissionGroupName>`

Reviewer 3: ```suggestion
When using additional permission groups, the permission name is derived from the `entryPointUriPath` (same as the default group) plus the group name, based on the following format: `{View,Manage}<EntryPointUriPath><GroupName>`
```
